### ISSUE TITLE
Poetry installing wrong version of pyglet (2.0dev2 instead of 2.0a2)

### ISSUE THREAD

**Issue by villekr** [pythonarcade/arcade#1156]
## System Info

MacOS 11.2.3 Big Sur
Python 3.10.2 (Using pyenv and Poetry virtualenv)
arcade 2.6.13

### Actual behavior:

  File ".../python3.10/site-packages/arcade/__init__.py", line 102, in <module>
    from .camera import Camera
  File ".../python3.10/site-packages/arcade/camera.py", line 7, in <module>
    from pyglet.math import Mat4, Vec2, Vec3

ImportError: cannot import name 'Vec2' from 'pyglet.math'

### Steps to reproduce/example code:

import arcade

### What should be added/changed?

arcade library pypi.org release's pyglet dependant version doesn't yet have imported Vec2 classes

arcade library setup.py has the following:
    install_requires=[
        "pyglet==2.0a2",

### What would it help with?

Update required pyglet version.

**Comment by einarf** [comment]
Can you check what versions of arcade and pyglet potery actually installed (`pip freeze`)? I've seen a couple of people having problems with potery installing older version. If you have `pyglet==2.0a2` and `arcade==2.6.13` this should work.

It's either some bug in poetry or we need to tweak our requirements a bit to make potery happy. If you create a clean virtualenv and install dependecies again with poetry, are there any useful output that would hint on what is happening? (If it goes wrong)

Could it be that poetry doesn't like arcade using pre-release of pyglet 2.0?

**Comment by einarf** [comment]
Just setting up everything manually seems to work fine

```bash
$ pyenv shell 3.10.2  // switch to 3.10.2
$ python -m venv .venv
$ source .venv/bin/activate
$ pip install arcade==2.6.13
$ python
>>> import arcade
>>>
```

`python -m pip freeze`:

```
arcade==2.6.13
attrs==21.4.0
cffi==1.15.0
Pillow==9.0.1
pycparser==2.21
pyglet==2.0a2
pymunk==6.2.1
pytiled-parser==2.0.1
typing_extensions==4.1.1
```

However with poetry it doesn't work for some odd reason. Poetry installs `pyglet-2.0.dev2-py3-none-any.whl` instead of 2.0a2.

```ini
[tool.poetry]
name = "poetry_test"
version = "0.1.0"
description = "Testing poetry"
authors = ["Your Name <you@example.com>"]

[tool.poetry.dependencies]
python = "^3.10"
arcade = "^2.6.13"

[tool.poetry.dev-dependencies]

[build-system]
requires = ["poetry-core>=1.0.0"]
build-backend = "poetry.core.masonry.api"
```

`python -  m pip freeze`:

```
arcade @ file:///C:/Users/efors/AppData/Local/pypoetry/Cache/artifacts/4a/8f/25/1cef8427287a21e6912e39cc2cd2da3fae1e89b187059b549e7fbea9c7/arcade-2.6.13-py3-none-any.whl
Pillow @ file:///C:/Users/efors/AppData/Local/pypoetry/Cache/artifacts/7c/10/f6/deba9b6ca3b3ca4beb43bfa6e5167a46f68835360a3c30a47a2a497a57/Pillow-9.0.1-cp310-
pyglet @ file:///C:/Users/efors/AppData/Local/pypoetry/Cache/artifacts/6e/fd/c0/e55bff14bb156b5dd8d452d955876aeeb19bb1e50bcab3543f4ccd0bf0/pyglet-2.0.dev2-py3-none-any.whl
pymunk @ file:///C:/Users/efors/AppData/Local/pypoetry/Cache/artifacts/7f/7c/72/63f716f6f43a0632fd3b0ad58c37f1d4b8cddfcc6fdf258ebee4aa9fb8/pymunk-6.2.1-cp310-cp310-win_amd64.whl  
```

I'm not a regular user of poetry, but this looks like a bug on their side? We specifically **pin** the exact version of pyglet, so I don't understand why it would change that. It's probably parsing the version and mixing up dev and alpha releases?

**Comment by einarf** [comment]
I'm guessing they follow this:
https://peps.python.org/pep-0440/#developmental-releases

`2.0devX` is valid, but `2.0aX` is not. Pyglet will change to `2.0.aX` in the next releases.

Still, it's weird that a pinned dependency gets ignored. There might be some option in poetry to bypass this?

**Comment by villekr** [comment]
Thanks for reacting quickly and correcting the issue to show to actual problem.

My question is that why arcade's final release depends on pyglet alpha release. Wouldn't it be the correct approach to release arcade as alpha until needed api changes are included in pyglet final release version?

**Comment by villekr** [comment]
Pinning version to arcade = "2.6.7" in pyproject.toml works for me and might be the easiest workaround for poetry users for now.

**Comment by einarf** [comment]
> My question is that why arcade's final release depends on pyglet alpha release. Wouldn't it be the correct approach to release arcade as alpha until needed api changes are included in pyglet final release version?

This is part of a 2+ year long journey trying to get both pyglet and arcade to support OpenGL 3.3+ core and OpenGL ES 3.1/2. Arcade had to skip ahead and make their own redering system. We currently only use window, events, text and media features from pyglet 2, but pyglets own rendering are not yet ready. They need to go slower since 2.0 is supposed to be a transition version mostly compatible with 1.5.

* In versions up to arcade 2.5 we used pyglet 1.5 with some context and window hacks to get GL 3.3.
* In arcade 2.6 we swapped to pyglet 2.0 beta/alpha using selected features we knew was working
* In arcade 2.7 we will use pyglet 2.0 stable release and support rendering both with arcade and pyglet
* In arcade 2.8 / pyglet 2.1 we will try to merge the rendering systems

Arcade 2.7 / pyglet 2.0 (stable) should mostly also be compatible with Raspberry PI 4.

> Pinning version to arcade = "2.6.7" in pyproject.toml works for me and might be the easiest workaround for poetry users for now.

I think you can get away with 2.6.9 and even a little higher since these also use 2.0devxx pyglet versions. My main concern is that those depend on an insecure version of Pillow (8.x). I know poetry can downgrade arcade if you have other packages depending on pillow 8. arcade 2.6.7 was also the last version supporting py36. Is that even related to this?

Another option is to fork arcade and see if you can loosen up the dependecies a bit and see if that makes a difference.

**Comment by einarf** [comment]
Pyglet changed back to dev releases skipping alpha. Their alpha names was actually in line with PEP 440, but PyPI changed the version string from 2.0.a2 to 2.0a2 on  upload for some reason. There is definitely something weird with pre-releases.

We will  use the new 2.0.dev14 in next release. If poetry can't resolve that then I don't really know if there is anything we can do.

**Comment by einarf** [comment]
Should be fixed in 2.6.14. Poetry should understand pyglet==2.0.dev16
